math/factorial/factorial_test.go

```diff
@@ -1,81 +1,75 @@
 // factorial_test.go
 // description: Test for calculating factorial
-// author(s) [red_byte](https://github.com/i-redbyte)
 // see factorial.go
 
 package factorial
 
 import "testing"
+import "fmt"
 
-func getTests() []struct {
-	name   string
-	n      int
-	result int
-} {
-	var tests = []struct {
-		name   string
-		n      int
-		result int
-	}{
-		{"5!", 5, 120},
-		{"3!", 3, 6},
-		{"6!", 6, 720},
-		{"12!", 12, 479001600},
-		{"1!", 1, 1},
-	}
-	return tests
-}
+type factorialFun func(int) (int, error)
 
-func TestBruteForceFactorial(t *testing.T) {
-	tests := getTests()
-	for _, tv := range tests {
-		t.Run(tv.name, func(t *testing.T) {
-			result := Iterative(tv.n)
-			if result != tv.result {
-				t.Errorf("Wrong result! Expected:%d, returned:%d ", tv.result, result)
-			}
-		})
+func implementations() map[string]factorialFun {
+	return map[string]factorialFun{
+		"Iterative": Iterative,
+		"Recursive": Recursive,
+		"UsingTree": UsingTree,
 	}
 }
 
-func TestRecursiveFactorial(t *testing.T) {
-	tests := getTests()
-	for _, tv := range tests {
-		t.Run(tv.name, func(t *testing.T) {
-			result := Recursive(tv.n)
-			if result != tv.result {
-				t.Errorf("Wrong result! Expected:%d, returned:%d ", tv.result, result)
-			}
-		})
+func testCases() []struct {
+	n        int
+	expected int
+} {
+	return []struct {
+		n        int
+		expected int
+	}{
+		{0, 1},
+		{1, 1},
+		{2, 2},
+		{3, 6},
+		{4, 24},
+		{5, 120},
+		{6, 720},
+		{7, 5040},
+		{8, 40320},
+		{9, 362880},
+		{10, 3628800},
+		{11, 39916800},
+		{12, 479001600},
 	}
 }
-
-func TestCalculateFactorialUseTree(t *testing.T) {
-	tests := getTests()
-	for _, tv := range tests {
-		t.Run(tv.name, func(t *testing.T) {
-			result := UsingTree(tv.n)
-			if result != tv.result {
-				t.Errorf("Wrong result! Expected:%d, returned:%d ", tv.result, result)
+func TestFactorial(t *testing.T) {
+	for implName, implFunction := range implementations() {
+		t.Run(implName+"_errors_for_negative_input", func(t *testing.T) {
+			_, error := implFunction(-1)
+			if error != InputMustBeNonnegative {
+				t.Errorf("No error captured for negative input")
 			}
 		})
+		for _, tc := range testCases() {
+			t.Run(fmt.Sprintf("%s_with_input_%d", implName, tc.n), func(t *testing.T) {
+				actual, err := implFunction(tc.n)
+				if err != nil {
+					t.Errorf("Unexpected error captured")
```

Comment:
```suggestion
					t.Errorf("unexpected error captured")
```

For some reason, quite a few of linting tools complain about capitalisation of errors (I have no personal preference but my IDE starts complaining 😂 It didn't used to before so I never caught these)

Reply:
I am not that familiar with Go ecosystem to have a strong opinion here.